### ISSUE TITLE
breaking behavior due to change to event delegation in phoenix_html.js

### ISSUE THREAD

**Issue by montague** [phoenixframework/phoenix_html#96]
caused by https://github.com/phoenixframework/phoenix_html/pull/77.

this is a pretty significant change of behavior that broke an app i've been working on--if the `event.target` is not the anchor element, then the form won't submit. in my case, i'm using bootstrap and my form links wrap glyphicons. i deployed this new code and suddenly none of my delete buttons worked anymore.

in my example, clicking the icon in the markup below will no longer cause the form to submit.
```
<form><a data-submit='parent'><span class='glyphicon'><span></a></form>
```

it was straightforward to write a workaround, but the change in behavior was surprising. i'd expect a delete link generated by phoenix_html to work as expected when wrapping child elements.

**Comment by josevalim** [comment]
@montague good catch, thank you! Can you please provide a fix?

**Comment by montague** [comment]
sure, i'll get something together tonight.

**Comment by montague** [comment]
@josevalim Fixing this is tricky, unless I'm missing something. The current implementation has the problem I described above. It also fires the callback for every click on the window object, which isn't optimal.

The original implementation didn't support elements that were dynamically added to the DOM.

It looks to me like doing something with `MutationObserver` might be a path to look down, but its IE support is only 11+. Seems that we want to support IE 9+?

A simple way forward would be to revert #77 and then add a utility method that users can call on dynamically added `a[data-submit=parent]` elements to attach the necessary callback.

Thoughts?

**Comment by montague** [comment]
@josevalim Another technique would be to walk up the tree looking at each of the `event.target`'s ancestors to see if it matches `a[data-submit=parent]`. I just talked to a buddy who knows much more js than I do and he said walking up the DOM is a totally acceptable and performant approach, even for deeply nested DOMs. I'm game to give that a shot. Unless I hear otherwise, I'll open a PR for discussion tomorrow morning (I'm ET, so ~10 hrs from now).

**Comment by josevalim** [comment]
Let's copy @Troush since he wrote the original code. Maybe he has ideas on how to improve it. I think traversing the tree until you find the parent form is an option but maybe there are more efficient ways of doing so.

**Comment by jimbeaudoin** [comment]
Same problem here, I suddenly can't log out from an app I'm working on. The form is not submitted anymore.

**Comment by josevalim** [comment]
@jimbeaudoin just to confirm, is the same root cause? Your links are being wrapped into something else and that is triggering the failure?

It is quite interesting because the previous code supposedly has the same bug, since both of them rely on `parentNode`.

**Comment by jimbeaudoin** [comment]
@josevalim My code looks like this: `<li><%= link "Sign Out", to: session_path(@conn, :delete, @current_user), method: "delete" %></li>`. It's a menu with a sub-menu `ul > li > ul > li`. The sign out link is in the sub-menu.

**Comment by josevalim** [comment]
@jimbeaudoin perfect, thank you.

**Comment by jimbeaudoin** [comment]
@josevalim I noticed this bug tonight when installing my app inside a VM for learning purpose. I'm accessing this VM by using an internal ip like `192.168.1.2:4200`. When I launch the same app directly on my laptop `localhost:4200`, I can log out. From the VM, an hashtag is added to the url and the form is not submitted. I'm going to analyse the deps on both machines if I can find something.

**Comment by josevalim** [comment]
@jimbeaudoin when you deploy do you keep the `mix.lock`? If so it is guarantee both machines should use the same `phoenix_html` version. Then the remaining question is in your `package.json` and if you are sure the phoenix_html javascript has been included in the final compiled javascript.

**Comment by jimbeaudoin** [comment]
@josevalim yes I keep the `mix.lock`. I don't deploy on the VM. It's a local virtual machine I run on my laptop and it's accessible on my internal network. I use VM as dev machines. I ssh to it and work from there. Like this I can run different version of Erlang and Elixir without problems for different applications (older and new).

I tried the same setup (laptop and VM) with an older app (Elrang 18/Elixir 1.2) and everything is working good. The log out works. Going to investigate and if I find something, I'm going to write what I found.

**Comment by josevalim** [comment]
@montague btw, this may be helpful: http://stackoverflow.com/questions/14234560/javascript-how-to-get-parent-element-by-selector

**Comment by jimbeaudoin** [comment]
@josevalim I now have the same problem on my laptop. I did a `rm -rf node_modules && deps` and reinstall with `mix deps.get && npm install` and now I can't log out anymore on both places.

**Comment by montague** [comment]
@josevalim Thanks, that's the technique my friend the javascript wizard recommended--keep the code as-is, but also walk up the DOM tree looking for a parent that matches `a[data-submit=parent]`.

I don't think previous version of the code had the same bug, since the event handler was attached directly to the `a[data-submit=parent]`element. So when a child element was clicked, the event bubbled up to the `a[data-submit=parent]` and all was handled correctly. In this case, the code only triggers the callback if the `a[data-submit=parent]` element itself is the event target.

I had a similar experience to @jimbeaudoin, in that i noticed the change only yesterday, even though I upgraded to phoenix 1.2 thirteen days ago.

**Comment by montague** [comment]
In the interest of getting this fixed, here's some code we can talk about: #97

**Comment by josevalim** [comment]
Thank you @montague! The code looks good to me. I assume you have also tried it locally and it worked? :D

If so, I will merge it asap. @jimbeaudoin, please let me know if it solves the issue for you too.

**Comment by montague** [comment]
@josevalim yep, worked for both my original code and a dynamically created element.

**Comment by jimbeaudoin** [comment]
@josevalim the patch doesn't work for me but the old code works:

```javascript
'use strict';

// Although ^=parent is not technically correct,
// we need to use it in order to get IE8 support.
var elements = document.querySelectorAll('[data-submit^=parent]');
var len = elements.length;

for (var i = 0; i < len; ++i) {
  elements[i].addEventListener('click', function (event) {
    var message = this.getAttribute("data-confirm");
    if (message === null || confirm(message)) {
      this.parentNode.submit();
    };
    event.preventDefault();
    return false;
  }, false);
}

;
```
Thanks!

**Comment by montague** [comment]
hmm... @jimbeaudoin can you send me a gist of the exact markup that's broken?

**Comment by jimbeaudoin** [comment]
@montague my code looks like this: (I removed some links and css classes for readability)

```
<nav>
  <ul>
    <li><%= link "home", to: page_path(@conn, :index) %></li>
  </ul>
  <ul>
    <li id="user-menu">
      <span class="item"><%= @current_user.username %></span>
      <ul id="sub-menu">
        <li><%= link "Sign Out", to: session_path(@conn, :delete, @current_user), method: "delete" %></li>
      </ul>
    </li>
  </ul>
</nav>
```
`sub-menu` is `display: none` by default. When I click on `user-menu` then I update the css with JavaScript to show `sub-menu`.

**Comment by montague** [comment]
can you paste the markup that gets generated when that template renders?

basically, load the web page, view the source, and then copy that bit? should be an `a` tag inside a `form` tag inside an `li` tag.

**Comment by montague** [comment]
we'll get to the bottom of this :)

**Comment by jimbeaudoin** [comment]
@montague yes it is

```html
<li>
  <form action="/sessions/72582909-1ea5-492c-9833-e1ce1c34f70b" class="link" method="post">

    <input name="_method" type="hidden" value="delete">
    <input name="_csrf_token" type="hidden" value="c0hTJTs1CSAoZXA1HzBdPw0YCyEPNgAA6gewJAgZBSCG4i0LDJHcWw==">

    <a data-submit="parent" href="#" rel="nofollow">Sign Out</a>
  
  </form>
</li>
```

**Comment by montague** [comment]
my minimal reproduction of your markup worked as expected. can you open up your `app.js` file that gets rendered and make sure the new code is in there? 

is your site live somewhere that i can poke around?

**Comment by montague** [comment]
@jimbeaudoin ^

**Comment by jimbeaudoin** [comment]
@montague the new code was there. It's the first thing I checked. I'm going to make a small public app to reproduce the bug I have, after my lunch.

**Comment by montague** [comment]
@jimbeaudoin that'd be awesome, thanks. let me know when it's available and i'll clone it and take a look.

**Comment by jimbeaudoin** [comment]
@montague it's weird because in this sample application: https://github.com/jimbeaudoin/html_testr it's working with no modification (phoenix_html from deps). The only difference between the two application is the private one has been updated to phoenix 1.2 and the public one has been generated with phoenix 1.2. Going to take a look further into the code why it's happening in my private app.

**Comment by jimbeaudoin** [comment]
@montague I was able to reproduced the bug. I put some code to close the sub menu when clicking on the html element and I need to use `event.stopPropagation()`. This is what is causing the bug for me. 

You can take a look here: https://github.com/jimbeaudoin/html_testr/blob/master/web/static/js/nav-bar.js#L18

You can't logout anymore from the application because of this.

To reproduce the bug:
- clone the repos https://github.com/jimbeaudoin/html_testr
- start the app and click sub-menu and then login
- click sub-menu again and click logout. The bug should happen.

**Comment by montague** [comment]
thanks, @jimbeaudoin. i'm looking now.

**Comment by montague** [comment]
@josevalim @jimbeaudoin 

The old way of doing it, by attaching the event handler directly to the `a[data-submit=parent]` element, worked for @jimbeaudoin because the handler fired before it hit the `event.stopPropagation()` code he used to close the sub menu (see https://github.com/jimbeaudoin/html_testr/blob/dfa3bef4adb8688ff68218de75b656bc9ead5f87/web/static/js/nav-bar.js#L18).

The new code, which depends on event delegation, requires the event to bubble up to the window. The benefit of this is all `a[data-parent=submit]` elements will behave as expected, regardless of when they were added to the DOM.

`event.stopPropagation()` basically means  "pretend that this event didn't happen", so no additional "click" handlers will fire. I think that this is exactly how we want the code to work--jQuery dealt with the same problem and works similarly, so I think it's expected.

@jimbeaudoin you could work around this limitation by removing the `event.stopPropagation()` call and instead adding a flag to the event object when it fires the `userMenu` click event--something like the code below:

```
let NavBar = {
  menuIsShown: false,

  init(element) {
    if(!element) { return; }
    let userMenu = document.getElementById("user-menu");
    let userSubMenu = document.getElementById("sub-menu");
    let appHtml = document.getElementById("app-html");

    userMenu.addEventListener("click", e => {
      if (this.menuIsShown === true) {
        userSubMenu.style.display = "none";
      }
      else {
        userSubMenu.style.display = "block";
      }
      this.menuIsShown = !this.menuIsShown;
      e.toggled = true
    }),

    appHtml.addEventListener("click", e => {
      if (e.toggled) return;
      if (this.menuIsShown === true) {
        userSubMenu.style.display = "none";
        this.menuIsShown = false;
      }
    })
  }
}

export default NavBar;
```

tldr; i think this works the way most people would expect.

**Comment by josevalim** [comment]
Thank you @montague and @jimbeaudoin . I agree with your assessment, I will release a new version with the new javascript changes. 
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by montague** [comment]
Happy to contribute :)

**Comment by Troush** [comment]
Hey, missed this discussion.
As I understand, that we are okay with this?

**Comment by montague** [comment]
@Troush yep, i'll close this out. thanks for checking in 👍
